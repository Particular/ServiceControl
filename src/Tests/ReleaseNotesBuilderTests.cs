namespace ReleaseNotesCompiler.Tests
{
    using System;
    using System.Linq;
    using ApprovalTests;
    using NUnit.Framework;
    using Octokit;

    [TestFixture]
    public class ReleaseNotesBuilderTests
    {
        [Test]
        public void NoCommitsNoIssues()
        {
            AcceptTest(0);
        }

        [Test]
        public void NoCommitsSomeIssues()
        {
            AcceptTest(0, CreateIssue(1, "Bug"), CreateIssue(2, "Feature"), CreateIssue(3, "Improvement"));
        }

        [Test]
        public void SomeCommitsNoIssues()
        {
            AcceptTest(5);
        }

        [Test]
        public void SomeCommitsSomeIssues()
        {
            AcceptTest(5, CreateIssue(1, "Bug"), CreateIssue(2, "Feature"), CreateIssue(3, "Improvement"));
        }

        [Test]
        public void SingularCommitsNoIssues()
        {
            AcceptTest(1);
        }

        [Test]
        public void SingularCommitsSomeIssues()
        {
            AcceptTest(1, CreateIssue(1, "Bug"), CreateIssue(2, "Feature"), CreateIssue(3, "Improvement"));
        }

        [Test]
        public void SingularCommitsSingularIssues()
        {
            AcceptTest(1, CreateIssue(1, "Bug"));
        }

        [Test]
        public void NoCommitsSingularIssues()
        {
            AcceptTest(0, CreateIssue(1, "Bug"));
        }

        [Test]
        public void SomeCommitsSingularIssues()
        {
            AcceptTest(5, CreateIssue(1, "Bug"));
        }

        static void AcceptTest(int commits, params Issue[] issues)
        {
            var fakeClient = new FakeGitHubClient();

            fakeClient.Milestones.Add(CreateMilestone("1.2.3"));

            fakeClient.NumberOfCommits = commits;

            foreach (var issue in issues)
            {
                fakeClient.Issues.Add(issue);
            }

            var builder = new ReleaseNotesBuilder(fakeClient, "TestUser", "FakeRepo", "1.2.3");

            var notes = builder.BuildReleaseNotes().Result;

            Approvals.Verify(notes);
        }


        static Milestone CreateMilestone(string version)
        {
            return new Milestone(new Uri("https://github.com/Particular/FakeRepo/issues?q=milestone%3A" + version), 0, ItemState.Open, version, String.Empty, null, 0, 0, DateTimeOffset.Now, null, null);
        }

        static Issue CreateIssue(int number, params string[] labels)
        {
            return new Issue(null,
                new Uri("http://example.com/" + number),
                null,
                null,
                number,
                ItemState.Open,
                "Issue " + number,
                "Some issue",
                null,
                null,
                labels.Select(x => new Label(null, x, null)).ToArray(),
                null,
                null,
                0,
                null,
                null,
                DateTimeOffset.Now,
                null,
                1,
                false,
                null);
        }
    }
}