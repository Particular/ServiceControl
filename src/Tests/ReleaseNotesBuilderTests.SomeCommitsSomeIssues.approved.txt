As part of this release we had [5 commits](https://github.com/TestUser/FakeRepo/commits/1.2.3) which resulted in [3 issues](https://github.com/FakeRepo/issues/issues?milestone=0&state=closed) being closed.


## Improvements
- [**3**](http://example.com/3) - Issue 3  

## Bug Fixes
- [**1**](http://example.com/1) - Issue 1  

## Where to get it
You can download this release from [website](http://particular.net/downloads).
